src/BuiltInTools/dotnet-watch/HotReloadDotNetWatcher.cs

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Watcher.Internal;
+using Microsoft.DotNet.Watcher.Tools;
+using Microsoft.Extensions.Tools.Internal;
+
+namespace Microsoft.DotNet.Watcher
+{
+    public class HotReloadDotNetWatcher : IAsyncDisposable
+    {
+        private readonly IReporter _reporter;
+        private readonly ProcessRunner _processRunner;
+        private readonly DotNetWatchOptions _dotNetWatchOptions;
+        private readonly IWatchFilter[] _filters;
+
+        public HotReloadDotNetWatcher(IReporter reporter, IFileSetFactory fileSetFactory, DotNetWatchOptions dotNetWatchOptions)
+        {
+            Ensure.NotNull(reporter, nameof(reporter));
+
+            _reporter = reporter;
+            _processRunner = new ProcessRunner(reporter);
+            _dotNetWatchOptions = dotNetWatchOptions;
+
+            _filters = new IWatchFilter[]
+            {
+                new MSBuildEvaluationFilter(fileSetFactory),
+                new DotNetBuildFilter(_processRunner, _reporter),
+                new LaunchBrowserFilter(_dotNetWatchOptions),
+            };
+        }
+
+        public async Task WatchAsync(DotNetWatchContext context, CancellationToken cancellationToken)
+        {
+            var cancelledTaskSource = new TaskCompletionSource();
+            cancellationToken.Register(state => ((TaskCompletionSource)state).TrySetResult(),
+                cancelledTaskSource);
+
+            var processSpec = context.ProcessSpec;
+
+            if (context.SuppressMSBuildIncrementalism)
+            {
+                _reporter.Verbose("MSBuild incremental optimizations suppressed.");
+            }
+
+            while (true)
+            {
+                context.Iteration++;
+
+                for (var i = 0; i < _filters.Length; i++)
+                {
+                    await _filters[i].ProcessAsync(context, cancellationToken);
+                }
+
+                // Reset for next run
+                context.RequiresMSBuildRevaluation = false;
+
+                processSpec.EnvironmentVariables["DOTNET_WATCH_ITERATION"] = (context.Iteration + 1).ToString(CultureInfo.InvariantCulture);
+
+                var fileSet = context.FileSet;
+                if (fileSet == null)
+                {
+                    _reporter.Error("Failed to find a list of files to watch");
+                    return;
+                }
+
+                if (!fileSet.Project.IsNetCoreApp60OrNewer())
+                {
+                    _reporter.Error($"Hot reload based watching is only supported in .NET 6.0 or newer apps. Update the project's launchSettings.json to disable this feature.");
```

Comment:
This seems bad. Are you saying that anyone who has a mix of projects (5.0, 6.0) will need to manually go into their launchsettings.json to disable it? Could it be a property in the csproj instead?

Reply:
The launching project must be a net6 project since the APIs to apply hot reload are only available in net6.

Reply:
Can we make this fail earlier for 5.0 and earlier?

Reply:
You have to explicitly modify the launchSettings to enable hot reload. So it's a bit unusual to get in a bad state to begin with. This is the earliest we know of their tfm.